File: go/arraysandhashing/productofarrayexceptself_test.go
package arraysandhashing

import (
	"reflect"
	"strconv"
	"testing"
)

func TestProductExceptSelf(t *testing.T) {
	testCases := []struct {
		input []int
		want  []int
	}{
		{
			input: []int{1, 2, 3, 4},
			want:  []int{24, 12, 8, 6},
		},
		{
			input: []int{-1, 1, 0, -3, 3},
			want:  []int{0, 0, 9, 0, 0},
		},
	}

	for i, testCase := range testCases {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			got := productExceptSelf(testCase.input)
			if !reflect.DeepEqual(got, testCase.want) {
				t.Errorf("got %v, want %v", got, testCase.want)
			}
		})
	}
}

File: ts/src/arrays-and-hashing/product-of-array-except-self.test.ts
import { test, expect } from "vitest";
import productExceptSelf from "./product-of-array-except-self";

test("product of array except self", () => {
  const testCases: Array<{ input: number[]; want: number[] }> = [
    {
      input: [1, 2, 3, 4],
      want: [24, 12, 8, 6],
    },
    {
      input: [-1, 1, 0, -3, 3],
      want: [0, 0, 9, 0, 0],
    },
  ];

  for (const testCase of testCases) {
    const got = productExceptSelf(testCase.input);
    expect(got).toEqual(testCase.want);
  }
});
